apiserver/common/leadership.go

```diff
@@ -103,47 +107,108 @@ func (a *leadershipPinningAPI) PinnedLeadership() (params.PinnedLeadershipResult
 	return result, nil
 }
 
-// TODO (manadart 2018-10-29): Rename the two methods below (and on the client
-// side) to be [Un]PinApplicationLeaders, and derive the list of applications
-// based on the authenticating entity.
-
-// PinMachineApplications pins leadership for applications represented by units
-// running on the auth'd machine.
-func (a *leadershipPinningAPI) PinMachineApplications() (params.PinApplicationsResults, error) {
+// PinApplicationLeaders pins leadership for applications based on the auth
+// tag provided.
+func (a *LeadershipPinning) PinApplicationLeaders() (params.PinApplicationsResults, error) {
 	if !a.authorizer.AuthMachineAgent() {
 		return params.PinApplicationsResults{}, ErrPerm
 	}
-	return a.pinMachineAppsOps(a.pinner.PinLeadership)
+
+	tag := a.authorizer.GetAuthTag()
+	switch tag.Kind() {
+	case names.MachineTagKind:
+		return a.pinMachineApplications()
+	default:
+		return params.PinApplicationsResults{}, ErrPerm
+	}
 }
 
-// UnpinMachineApplications unpins leadership for applications represented by
-// units running on the auth'd machine.
-func (a *leadershipPinningAPI) UnpinMachineApplications() (params.PinApplicationsResults, error) {
+// UnpinApplicationLeaders unpins leadership for applications based on the auth
+// tag provided.
+func (a *LeadershipPinning) UnpinApplicationLeaders() (params.PinApplicationsResults, error) {
 	if !a.authorizer.AuthMachineAgent() {
 		return params.PinApplicationsResults{}, ErrPerm
 	}
-	return a.pinMachineAppsOps(a.pinner.UnpinLeadership)
+
+	tag := a.authorizer.GetAuthTag()
+	switch tag.Kind() {
+	case names.MachineTagKind:
+		return a.unpinMachineApplications()
+	default:
+		return params.PinApplicationsResults{}, ErrPerm
+	}
 }
 
-// pinMachineAppsOps runs the input pin/unpin operation against all
-// applications represented by units on the authorised machine.
-// An assumption is made that the validity of the auth tag has been verified
-// by the caller.
-func (a *leadershipPinningAPI) pinMachineAppsOps(op func(string, string) error) (params.PinApplicationsResults, error) {
-	result := params.PinApplicationsResults{}
+// GetMachineApplicationNames returns the applications associated with a
+// machine.
+func (a *LeadershipPinning) GetMachineApplicationNames() ([]string, error) {
```

Comment:
Assuming this is exported so that the machine manager API can get applications using this facility, it will need to accept the machine ID rather than attempt to get it from the auth tag.

Reply:
I'm unsure why, as you need the tag for `pinMachineAppsOps`, or at least you need the entity, converting to string seems like we're losing the type early on in the process. I'll update what I have, but I'm unsure that using `string` everywhere is such a wise idea.